### PR TITLE
Function tool callback

### PR DESCRIPTION
**Pull request by JoseLuckmann**
# Description

This is a feature that allows applying some human-in-the-loop concepts in FunctionTool.

Basically, a callback function is added that enables the developer to request user input in the middle of an agent interaction, as well as allowing any programmatic action.

## New Package?

Did I fill in the `tool.llamahub` section in the `pyproject.toml` and provide a detailed README.md for my new integration or package?

- [ ] Yes
- [X] No

## Version Bump?

Did I bump the version in the `pyproject.toml` file of the package I am updating? (Except for the `llama-index-core` package)

- [ ] Yes
- [X] No

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Your pull-request will likely not be merged unless it is covered by some form of impactful unit testing.

- [ ] I added new unit tests to cover this change
- [X] I believe this change is already covered by existing unit tests

## Suggested Checklist:

- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [ ] I have added Google Colab support for the newly added notebooks.
- [X] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] I ran `make format; make lint` to appease the lint gods

### CODE REVIEW

File: llama-index-core/llama_index/core/tools/function_tool.py [hunk 1]
```diff
@@ -109,19 +117,27 @@ def async_fn(self) -> AsyncCallable:
...
             raw_output=tool_output,
         )
 
     async def acall(self, *args: Any, **kwargs: Any) -> ToolOutput:
-        """Call."""
-        tool_output = await self._async_fn(*args, **kwargs)
+        """Async Call."""
+        tool_output = self._fn(*args, **kwargs)        
+        final_output_content = str(tool_output)
+        callback_output = self._run_callback(tool_output)
```
Comment (logan-markewich) [inline llama-index-core/llama_index/core/tools/function_tool.py:136]: Probably the callback should be async? Otherwise this will block the event loop (probably not ideal)

Comment (JoseLuckmann) [reply llama-index-core/llama_index/core/tools/function_tool.py:136]: Hmm, in our case we implemented this to allow requesting user confirmation and in this scenario the event loop remains blocked until the user confirms. If implementing it this way is not feasible, do you have any other suggestions?

Comment (JoseLuckmann) [reply llama-index-core/llama_index/core/tools/function_tool.py:136]: In my view, yes. We developed this feature so that when a FunctionTool is called, it can request user input that will influence the result or execution of that function.

That said, it makes sense for it not to be asynchronous. However, in our case, we use synchronous calls. If you believe it should be handled asynchronously, I can change it without any problems.

Comment (JoseLuckmann) [reply llama-index-core/llama_index/core/tools/function_tool.py:136]: Stopping to evaluate this point, the person who uses an asynchronous call in this case really doesn't make sense conceptually to stop the application.
So whoever uses the asynchronous callback in the asynchronous call will be in a different situation.
I will make the adjustment

**Comment by JoseLuckmann** [comment]
Hello @logan-markewich, can you check it out? I've done the asynchronous implementation and now everything is up to standard

File: llama-index-core/llama_index/core/tools/function_tool.py [hunk 2]
```diff
@@ -108,20 +144,30 @@ def async_fn(self) -> AsyncCallable:
...
 
     def call(self, *args: Any, **kwargs: Any) -> ToolOutput:
-        """Call."""
+        """Sync Call."""
         tool_output = self._fn(*args, **kwargs)
+        final_output_content = str(tool_output)
+        # Execute sync callback, if available
+        callback_output = self._run_sync_callback(tool_output)
+        if callback_output:
+            final_output_content += f" Callback: {callback_output}"
```
Comment (logan-markewich) [inline llama-index-core/llama_index/core/tools/function_tool.py:153]: Actually, looking at this again, this seems rather opinionated 😅 What's the intended use case for this? 

I kind of expected the callback to just either be logging something, or outright modifying the tool output in place.

Appending a string like this seems kind of strange

Comment (JoseLuckmann) [reply llama-index-core/llama_index/core/tools/function_tool.py:153]: I'll explain the use case we applied here to see if it becomes a bit clearer:

I have a tool that sends information to the external system of my company.
The agent is responsible for setting all parameters for the API call, but I can't blindly trust the information constructed by the agent.
So, I use this callback strategy to request user confirmation.
This confirmation returns to the agent's interaction, allowing the agent to decide the next steps.
Here, we apply this rule with interactions between agents and the front end for confirmation, but if you check the example notebook I attached, you can see this flow abstracted in a certain way that's easier to understand.

Basically, I need the callback return to influence the remainder of the agent's flow. I couldn't think of an easier way to adapt the classes without violating any principles. If the user doesn't want the callback to influence the flow, they just don't return anything in the function.

Comment (logan-markewich) [reply llama-index-core/llama_index/core/tools/function_tool.py:153]: It seems like to me a better design might be 

```
callback_output: ToolOutput | None = self._run_sync_callback(tool_output)
```

Basically the callback either returns a new tool output to override the existing, or returns None and the original tool_output is used

Comment (logan-markewich) [reply llama-index-core/llama_index/core/tools/function_tool.py:153]: Then its up to the user how the callback changes the tool output

Comment (JoseLuckmann) [reply llama-index-core/llama_index/core/tools/function_tool.py]: Hello, sorry for the delay in making the adjustments.

I made the adjustments as requested, thinking of giving the user autonomy to choose whether he wants the callback output to be inserted into the tool's output.

**Comment by JoseLuckmann** [comment]
Hello @logan-markewich, could you take a look to understand if this MR is feasible
I am currently maintaining a customized version of llamaindex in my product to be able to implement the feature and would like to update

**Review by logan-markewich — APPROVED** [review]

**Comment by logan-markewich** [comment]
Sorry for the delay, seems good, will be in the next release
